src/porepy/numerics/ad/operator_functions.py

```diff
@@ -0,0 +1,463 @@
+        # the function is passed during a call to the decorator as first argument
+        if self._func is None:
+            self._func = args[0]
+
+        # if an explicit init was made,
+        # mimic a non-explicit init to get an object with descriptor protocol
+        if self._explicit_init:
+            # TODO VL: check if the ADmethod instance whose __call__ is executed right now
+            # gets properly de-referenced and deleted, or if it remains hidden in the memory
+            return ADmethod(
+                func=self._func,
+                ad_function_type=self._ad_func_type,
+                operator_kwargs=self._op_kwargs,
+            )
+
+        # without an explicit init, the first decorator itself replaces the decorated function
+        # This results in a call to ADmethod.__call__ instead of
+        # a call to the decorated function
+
+        # when calling the decorator, distinguish between bound method call
+        # ('args' contains 'self' of the decorated instance) and an unbound function call
+        # (whatever 'args' and 'kwargs' contain, we pass it to the wrapper)
+        if self._bound_to is None:
+            wrapped_function = self.ad_wrapper(*args, **kwargs)
+        elif self._bound_to == args[0]:
+            wrapped_function = self.ad_wrapper(*args[1:], **kwargs)
+        else:
+            raise ValueError(
+                "Calling bound decorator "
+                + str(self)
+                + " with unknown instance "
+                + str(args[0])
+                + "\n This decorator is bound to "
+                + str(self._bound_to)
+            )
+
+        return wrapped_function
+
+    def ad_wrapper(self, *args, **kwargs):
+        """
+        Actual wrapper function.
+        Constructs the necessary AD-Operator class and performs the evaluation.
+        """
+        # extra safety measure to ensure a a bound call is done to the right, binding instance.
+        # We pass only the binding instance referenced in the descriptor protocol.
+        if self._bound_to is None:
+            operator_func = self._func
+        else:
+            # partial evaluation of a bound function,
+            # since the AD operator has no reference to binding instance
+            operator_func = partial(self._func, self._bound_to)
+
+        # Resulting AD operator has a special name to mark its origin
+        if "name" not in self._op_kwargs.keys():
+            name = "ADmethod-"
+            name += (
+                str(self._ad_func_type.__qualname__)
+                + "-decorating-"
+                + str(self._func.__qualname__)
+            )
+            self._op_kwargs.update({"name": name})
+
+        # calling the operator
+        wrapping_operator = self._ad_func_type(func=operator_func, **self._op_kwargs)
+
+        return wrapping_operator(*args, **kwargs)
+
+    def __get__(self, binding_instance: object, binding_type: type):
```

Comment:
I don't understand this at all. A much simpler documentation (less technical) documentation is needed.

Reply:
Unfortunately, this is the best I can do and it is a complex issue.
This is all due to the problem, that I want this class to be able to decorate also class methods, not just plain functions.
See my comment two conversations above.